File: api/ticker.py
from http.server import BaseHTTPRequestHandler
import urllib.request
import json

class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        try:
            # Fetch Crypto Data from CoinGecko
            req_url = 'https://api.coingecko.com/api/v3/simple/price?ids=bitcoin,ethereum,solana,tether,ripple,cardano,dogecoin,polkadot&vs_currencies=inr,usd&include_24hr_change=true'
            req = urllib.request.Request(req_url, headers={'User-Agent': 'Mozilla/5.0'})
            
            with urllib.request.urlopen(req) as response:
                data = json.loads(response.read().decode())
            
            coins = [
                {'id': 'bitcoin', 'symbol': 'BTC'},
                {'id': 'ethereum', 'symbol': 'ETH'},
                {'id': 'solana', 'symbol': 'SOL'},
                {'id': 'tether', 'symbol': 'USDT'},
                {'id': 'ripple', 'symbol': 'XRP'},
                {'id': 'cardano', 'symbol': 'ADA'},
                {'id': 'dogecoin', 'symbol': 'DOGE'},
                {'id': 'polkadot', 'symbol': 'DOT'}
            ]
            
            compiled_data = []
            for coin in coins:
                coin_data = data.get(coin['id'])
                if coin_data:
                    compiled_data.append({
                        'id': coin['id'],
                        'symbol': coin['symbol'],
                        'priceUsd': coin_data.get('usd'),
                        'priceInr': coin_data.get('inr'),
                        'changePercent24Hr': coin_data.get('usd_24h_change')
                    })
                
            self.send_response(200)
            self.send_header('Content-type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.send_header('Cache-Control', 's-maxage=60, stale-while-revalidate=120')
            self.end_headers()
            self.wfile.write(json.dumps({'data': compiled_data}).encode())
            
        except Exception as e:
            self.send_response(500)
            self.send_header('Content-type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            self.wfile.write(json.dumps({'error': str(e)}).encode())

File: api/ticker.js
export const config = { runtime: 'edge' };

export default async function handler(req) {
  try {
    const response = await fetch(
      'https://api.coingecko.com/api/v3/simple/price?ids=bitcoin,ethereum,solana,tether,ripple,cardano,dogecoin,polkadot&vs_currencies=inr,usd&include_24hr_change=true',
      { headers: { 'User-Agent': 'Mozilla/5.0' }, next: { revalidate: 10 } }
    );
    
    if (!response.ok) throw new Error(`CoinGecko status: ${response.status}`);
    const data = await response.json();
    
    const RETAIL_MARKUP = 1.10; // 10% retail markup accounting for 1% TDS (Sec 194S) and platform spread

    const coins = [
      { id: 'bitcoin', symbol: 'BTC' },
      { id: 'ethereum', symbol: 'ETH' },
      { id: 'solana', symbol: 'SOL' },
      { id: 'tether', symbol: 'USDT' },
      { id: 'ripple', symbol: 'XRP' },
      { id: 'cardano', symbol: 'ADA' },
      { id: 'dogecoin', symbol: 'DOGE' },
      { id: 'polkadot', symbol: 'DOT' }
    ];
    
    const compiledData = coins.map(coin => ({
      id: coin.id,
      symbol: coin.symbol,
      priceUsd: (data[coin.id]?.usd || 0) * RETAIL_MARKUP,
      priceInr: (data[coin.id]?.inr || 0) * RETAIL_MARKUP,
      changePercent24Hr: data[coin.id]?.usd_24h_change || 0
    }));

    return new Response(JSON.stringify({ data: compiledData }), {
      status: 200,
      headers: {
        'Content-Type': 'application/json',
        'Access-Control-Allow-Origin': '*',
        'Cache-Control': 'public, s-maxage=10, stale-while-revalidate=10'
      }
    });
  } catch (err) {
    return new Response(JSON.stringify({ error: err.message }), { status: 500 });
  }
}
